mathphys/imgproc.py

```diff
@@ -0,0 +1,1343 @@
+            fig, axis = _plt.subplots()
+
+        color = [0, 0.7, 0]
+
+        axis.plot(
+            self.roi_indcs, self.roi_proj,
+            color=color, alpha=1.0,
+            lw=5, label='roi_proj')
+        vecy, vecx = self.roi_fit
+        if vecy is not None:
+            axis.plot(
+                vecx, vecy, color=[0.5, 1, 0.5], alpha=1.0,
+                lw=2, label='roix_fit')
+
+        axis.legend()
+        axis.grid()
+        axis.set_ylabel('ROI pixel index')
+        axis.set_ylabel('Projection Intensity')
+
+    def __str__(self):
+        """."""
+        res = super().__str__()
+        res += f'\nroi_amplitude   : {self.roi_amplitude}'
+        res += f'\nroi_mean        : {self.roi_mean}'
+        res += f'\nroi_sigma       : {self.roi_sigma}'
+        res += f'\nroi_fit_err     : {self.roi_fit_error} %'
+
+        return res
+
+    def _update_image_roi(self, roi=None, *args, **kwargs):
+        """."""
+        super()._update_image_roi(roi=roi)
+
+        # fit roi
+        param, roi_fit, roi_error = \
+            self._fitgauss.calc_fit(
+                self, self.roi_proj, self.roi_indcs, self.roi_center)
+        self._roi_sigma, self._roi_mean, self._roi_amp, _ = param
+        self._roi_fit, self._roi_fit_error = roi_fit, roi_error
+
+
+class Image2D_Fit(Image2D):
+    """2D Image Fit."""
+
+    def __init__(self, roix=None, roiy=None, fitgauss=None, *args, **kwargs):
+        """."""
+        # benchmark for sizes=(1024, 1280)
+        #   2.7 ms ± 23.2 µs per loop
+        #   (mean ± std. dev. of 7 runs, 100 loops each)
+        self._fitx = None
+        self._fity = None
+        self._angle = 0
+        self._fitgauss = fitgauss or FitGaussian
+        super().__init__(*args, **kwargs)
+        self._update_image_fit(roix=roix, roiy=roiy)
+
+    @property
+    def fity(self):
+        """."""
+        return self._fity
+
+    @property
+    def fitx(self):
+        """."""
+        return self._fitx
+
+    @property
+    def roiy(self):
+        """."""
+        return self.fity.roi
+
+    @roiy.setter
+    def roiy(self, value):
+        """."""
+        self._update_image_fit(roix=self.fitx.roi, roiy=value)
+
+    @property
+    def roix(self):
+        """."""
+        return self.fitx.roi
+
+    @roix.setter
+    def roix(self, value):
+        """."""
+        self._update_image_fit(roix=value, roiy=self.fity.roi)
+
+    @property
+    def roi(self):
+        """."""
+        return self.fitx.roi, self.fity.roi
+
+    @roi.setter
+    def roi(self, value):
+        """."""
+        self._update_image_fit(*value)
+
+    @property
+    def angle(self):
+        """Image tilt angles [deg]."""
+        return self._angle
+
+    def update_roi_with_fwhm(self, fwhmx_factor=2, fwhmy_factor=2):
+        """."""
+        self.fitx.update_roi_with_fwhm(fwhm_factor=fwhmx_factor)
+        self.fity.update_roi_with_fwhm(fwhm_factor=fwhmy_factor)
+
+    def calc_angle_with_roi(self, sigma_factor, nrpts):
```

Comment:
What is happening in this method? Can we talk tomorrow?

Reply:
I think I understood now, but isn't there a better way of doing this? Maybe fitting a straight line to all points considering they are weighted by the intensity?

Reply:
I will try it, as we discussed